# Импорт библиотек

In [45]:
import pandas as pd
import numpy as np
import plotly.express as px
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, Ridge
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.preprocessing import PolynomialFeatures

# 1)

In [4]:
df = pd.read_csv('/content/boston.csv')
df

,CRIM,ZN,INDUS,CHAS,NOX,RM,AGE,DIS,RAD,TAX,PTRATIO,B,LSTAT,MEDV
0,0.00632,18.0,2.31,0.0,0.538,6.575,65.2,4.0900,1.0,296.0,15.3,396.90,4.98,24.0
1,0.02731,0.0,7.07,0.0,0.469,6.421,78.9,4.9671,2.0,242.0,17.8,396.90,9.14,21.6
2,0.02729,0.0,7.07,0.0,0.469,7.185,61.1,4.9671,2.0,242.0,17.8,392.83,4.03,34.7
3,0.03237,0.0,2.18,0.0,0.458,6.998,45.8,6.0622,3.0,222.0,18.7,394.63,2.94,33.4
4,0.06905,0.0,2.18,0.0,0.458,7.147,54.2,6.0622,3.0,222.0,18.7,396.90,5.33,36.2
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
501,0.06263,0.0,11.93,0.0,0.573,6.593,69.1,2.4786,1.0,273.0,21.0,391.99,9.67,22.4
502,0.04527,0.0,11.93,0.0,0.573,6.120,76.7,2.2875,1.0,273.0,21.0,396.90,9.08,20.6
503,0.06076,0.0,11.93,0.0,0.573,6.976,91.0,2.1675,1.0,273.0,21.0,396.90,5.64,23.9
504,0.10959,0.0,11.93,0.0,0.573,6.794,89.3,2.3889,1.0,273.0,21.0,393.45,6.48,22.0


# 2)

In [5]:
display(df.dtypes)
print('\nДанные имеют числовой тип')

,0
CRIM,float64
ZN,float64
INDUS,float64
CHAS,float64
NOX,float64
RM,float64
AGE,float64
DIS,float64
RAD,float64
TAX,float64



Данные имеют числовой тип


# 3)

In [6]:
display(df.isna().sum())
print("\nОтсутствуют пропущенные значения")

,0
CRIM,0
ZN,0
INDUS,0
CHAS,0
NOX,0
RM,0
AGE,0
DIS,0
RAD,0
TAX,0



Отсутствуют пропущенные значения


# 4)

In [7]:
corr_df = df.corr()

print('Матрица корелляции:')
display(corr_df)

Матрица корелляции:


,CRIM,ZN,INDUS,CHAS,NOX,RM,AGE,DIS,RAD,TAX,PTRATIO,B,LSTAT,MEDV
CRIM,1.000000,-0.200469,0.406583,-0.055892,0.420972,-0.219247,0.352734,-0.379670,0.625505,0.582764,0.289946,-0.385064,0.455621,-0.388305
ZN,-0.200469,1.000000,-0.533828,-0.042697,-0.516604,0.311991,-0.569537,0.664408,-0.311948,-0.314563,-0.391679,0.175520,-0.412995,0.360445
INDUS,0.406583,-0.533828,1.000000,0.062938,0.763651,-0.391676,0.644779,-0.708027,0.595129,0.720760,0.383248,-0.356977,0.603800,-0.483725
CHAS,-0.055892,-0.042697,0.062938,1.000000,0.091203,0.091251,0.086518,-0.099176,-0.007368,-0.035587,-0.121515,0.048788,-0.053929,0.175260
NOX,0.420972,-0.516604,0.763651,0.091203,1.000000,-0.302188,0.731470,-0.769230,0.611441,0.668023,0.188933,-0.380051,0.590879,-0.427321
RM,-0.219247,0.311991,-0.391676,0.091251,-0.302188,1.000000,-0.240265,0.205246,-0.209847,-0.292048,-0.355501,0.128069,-0.613808,0.695360
AGE,0.352734,-0.569537,0.644779,0.086518,0.731470,-0.240265,1.000000,-0.747881,0.456022,0.506456,0.261515,-0.273534,0.602339,-0.376955
DIS,-0.379670,0.664408,-0.708027,-0.099176,-0.769230,0.205246,-0.747881,1.000000,-0.494588,-0.534432,-0.232471,0.291512,-0.496996,0.249929
RAD,0.625505,-0.311948,0.595129,-0.007368,0.611441,-0.209847,0.456022,-0.494588,1.000000,0.910228,0.464741,-0.444413,0.488676,-0.381626
TAX,0.582764,-0.314563,0.720760,-0.035587,0.668023,-0.292048,0.506456,-0.534432,0.910228,1.000000,0.460853,-0.441808,0.543993,-0.468536


# 5)

In [8]:
fig5 = px.imshow(corr_df)
fig5.update_layout(width=777,height=777)
print('График матрицы корреляции:')
fig5.show()

График матрицы корреляции:


# 6)

In [9]:
most_corr = corr_df['MEDV'].abs().sort_values(ascending=False)
top5_features = most_corr[1:6]
top5_features_list = top5_features.index.tolist()
print("5 признаков которые наиболее коррелируют с ценной недвижимости:")
display(top5_features)

5 признаков которые наиболее коррелируют с ценной недвижимости:


,MEDV
LSTAT,0.737663
RM,0.695360
PTRATIO,0.507787
INDUS,0.483725
TAX,0.468536


# 7)

In [10]:
print('Матрица диаграмм рассеяния для выбранных признаков:')
for col in df[top5_features_list]:
  fig7 = px.scatter(df, x='MEDV', y=col, title=f"MEDV and {col}")
  fig7.show()

Матрица диаграмм рассеяния для выбранных признаков:


# 8)

* **MEDV** and **LSTAT**: Сильная отрицательная зависимость. Чем меньше процент населения с низким социальным статусом, тем выше цена недвижимости.

* **MEDV** and **RM**: Сильная положительная зависимость. Чем больше кол-во комнат в недвижимости, тем больше ее цена.

* **MEDV** and **PTRATIO**: Малая отрицательная зависимость. Чем меньше соотношение кол-ва учеников и учителей, тем больше недвижимости.

* **MEDV** and **INDUS**: Малая отрицательная зависимость. Чем меньше доля площадей, не связанных с розничной торговлей, тем выше цена недвижимости.

* **MEDV** and **TAX**: Малая отрицательная зависимость. Чем меньше налоги, тем выше цена недвижимости.

# 9)

In [15]:
X = df[top5_features_list]
y = df['MEDV']
display(x)
display(y)

,LSTAT,RM,PTRATIO,INDUS,TAX
0,4.98,6.575,15.3,2.31,296.0
1,9.14,6.421,17.8,7.07,242.0
2,4.03,7.185,17.8,7.07,242.0
3,2.94,6.998,18.7,2.18,222.0
4,5.33,7.147,18.7,2.18,222.0
...,...,...,...,...,...
501,9.67,6.593,21.0,11.93,273.0
502,9.08,6.120,21.0,11.93,273.0
503,5.64,6.976,21.0,11.93,273.0
504,6.48,6.794,21.0,11.93,273.0


,MEDV
0,24.0
1,21.6
2,34.7
3,33.4
4,36.2
...,...
501,22.4
502,20.6
503,23.9
504,22.0


# 10)

In [16]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print(f"Кол-во строк данных в train выбоорке: {X_train.shape[0]}")
print(f"Кол-во строк данных в test выбоорке: {X_test.shape[0]}")

Кол-во строк данных в train выбоорке: 404
Кол-во строк данных в test выбоорке: 102


# 11)

In [19]:
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

# 12)

In [21]:
y_train_prediction = model.predict(X_train)
y_test_prediction = model.predict(X_test)

# 13)

In [25]:
r2_train_score = r2_score(y_train, y_train_prediction)
rmse_train_score = np.sqrt(mean_squared_error(y_train, y_train_prediction))
r2_test_score = r2_score(y_test, y_test_prediction)
rmse_test_score = np.sqrt(mean_squared_error(y_test, y_test_prediction))

print(f"Значение R^2 ошибки на тренировочную выборку: {r2_train_score}")
print(f"Значение RMSE ошибки на тренировочную выборку: {rmse_train_score}")
print(f"Значение R^2 ошибки на тестовую выборку: {r2_test_score}")
print(f"Значение RMSE ошибки на тестовую выборку: {rmse_test_score}")

Значени R^2 ошибки на тренировочную выборку: 0.6902578367787746
Значени RMSE ошибки на тренировочную выборку: 5.187326481879791
Значени R^2 ошибки на тестовую выборку: 0.6196683547405641
Значени RMSE ошибки на тестовую выборку: 5.281208043853636


# 14)

In [29]:
fig14 = px.box(x=df['MEDV'])
fig14.update_layout(width=777, height=444, title='Разброс MEDV', xaxis_title='Цена за недижимсть')
fig14.show()
print("\nМожно заметить что присутствуют выбросы начиная со значения 37")


Можно заметить что присутствуют выбросы начиная со значения 37


# 15)

In [56]:
df_wo_outliers = df[df['MEDV'] < 37]

X_f = df_wo_outliers[top5_features_list]
y_f = df_wo_outliers['MEDV']

X_train_f, X_test_f, y_train_f, y_test_f = train_test_split(X_f, y_f, test_size=0.2, random_state=42)

model_f = LinearRegression()
model_f.fit(X_train_f, y_train_f)

y_train_prediction_f = model_f.predict(X_train_f)
y_test_prediction_f = model_f.predict(X_test_f)

r2_train_score_f = r2_score(y_train_f, y_train_prediction_f)
rmse_train_score_f = np.sqrt(mean_squared_error(y_train_f, y_train_prediction_f))
r2_test_score_f = r2_score(y_test_f, y_test_prediction_f)
rmse_test_score_f = np.sqrt(mean_squared_error(y_test_f, y_test_prediction_f))

print("Значения ошибок на данные без выбросов в признаке MEDV")
print(f"Значение R^2 ошибки на тренировочную выборку: {r2_train_score_f}")
print(f"Значение RMSE ошибки на тренировочную выборку: {rmse_train_score_f}")
print(f"Значение R^2 ошибки на тестовую выборку: {r2_test_score_f}")
print(f"Значение RMSE ошибки на тестовую выборку: {rmse_test_score_f}")

Значения ошибок на данные без выбросов в признаке MEDV
Значение R^2 ошибки на тренировочную выборку: 0.710042283701513
Значение RMSE ошибки на тренировочную выборку: 3.6118415175964484
Значение R^2 ошибки на тестовую выборку: 0.6888421262134915
Значение RMSE ошибки на тестовую выборку: 3.159148973339381


Сравнения:
* R2 ошибка увеличилась c 0.6196 до 0.6888
* RMSE ошибка уменьшилась с 5.2812 до 3.1591

Вывод:
* Линейная регрессия лучше работает с данными без выбросов.
* Порой лучше избавиться от выбросов при обучении модели.

# 16)

In [55]:
model_f_r = Ridge(alpha=1.0)
model_f_r.fit(X_train_f, y_train_f)

y_train_prediction_f_r = model_f_r.predict(X_train_f)
y_test_prediction_f_r = model_f_r.predict(X_test_f)

r2_train_score_f_r = r2_score(y_train_f, y_train_prediction_f_r)
rmse_train_score_f_r = np.sqrt(mean_squared_error(y_train_f, y_train_prediction_f_r))
r2_test_score_f_r = r2_score(y_test_f, y_test_prediction_f_r)
rmse_test_score_f_r = np.sqrt(mean_squared_error(y_test_f, y_test_prediction_f_r))

print("Значения ошибок на данные без выбросов в признаке MEDV и с использованием модели Ridge regression")
print(f"Значение R^2 ошибки на тренировочную выборку: {r2_train_score_f_r}")
print(f"Значение RMSE ошибки на тренировочную выборку: {rmse_train_score_f_r}")
print(f"Значение R^2 ошибки на тестовую выборку: {r2_test_score_f_r}")
print(f"Значение RMSE ошибки на тестовую выборку: {rmse_test_score_f_r}")

Значения ошибок на данные без выбросов в признаке MEDV и с использованием модели Ridge regression
Значение R^2 ошибки на тренировочную выборку: 0.7100388602091685
Значение RMSE ошибки на тренировочную выборку: 3.6118628398006725
Значение R^2 ошибки на тестовую выборку: 0.6890050474699807
Значение RMSE ошибки на тестовую выборку: 3.158321804956355


Сравнения:
* R2 ошибка увеличилась c 0.6888 до 0.689
* RMSE ошибка уменьшилась с 3.1591 до 3.1583

Вывод:
* Мультиколлинеарность и переобучение не является проблемой для линейной регрессии на этих данных.

# 17)

In [57]:
poly = PolynomialFeatures(degree=3)

X_poly_train = poly.fit_transform(X_train_f)
X_poly_test = poly.fit_transform(X_test_f)

model_poly = LinearRegression()
model_poly.fit(X_poly_train, y_train_f)

y_train_prediction_f_p = model_poly.predict(X_poly_train)
y_test_prediction_f_p = model_poly.predict(X_poly_test)

r2_train_score_f_p = r2_score(y_train_f, y_train_prediction_f_p)
rmse_train_score_f_p = np.sqrt(mean_squared_error(y_train_f, y_train_prediction_f_p))
r2_test_score_f_p = r2_score(y_test_f, y_test_prediction_f_p)
rmse_test_score_f_p = np.sqrt(mean_squared_error(y_test_f, y_test_prediction_f_p))

print("Значения ошибок на данные без выбросов в признаке MEDV и с использованием полиномиальной регрессии 3-ей степени:")
print(f"Значение R^2 ошибки на тренировочную выборку: {r2_train_score_f_p}")
print(f"Значение RMSE ошибки на тренировочную выборку: {rmse_train_score_f_p}")
print(f"Значение R^2 ошибки на тестовую выборку: {r2_test_score_f_p}")
print(f"Значение RMSE ошибки на тестовую выборку: {rmse_test_score_f_p}")

Значения ошибок на данные без выбросов в признаке MEDV и с использованием полиномиальной регрессии 3-ей степени:
Значение R^2 ошибки на тренировочную выборку: 0.8453816042180283
Значение RMSE ошибки на тренировочную выборку: 2.637496743266122
Значение R^2 ошибки на тестовую выборку: 0.8449043557936169
Значение RMSE ошибки на тестовую выборку: 2.2303833221344926


In [59]:
df['MEDV'].describe()

,MEDV
count,506.000000
mean,22.532806
std,9.197104
min,5.000000
25%,17.025000
50%,21.200000
75%,25.000000
max,50.000000


Выод по трем моделям:
* Сраниая по R^2 ошибке лучший результат показала модель полиномиальной регрессии 3-ей степени: 0.8449. Модель значительно лучше чем предскзаание средним, так как значение ошибки близко к единице.
* Сранивая по RMSE ошибке значительно лучший результат показала модель полиномиальной регрессии 3-ей степени: 2.2303. Модель значительно лучше чем предскзаание средним, так как значение ошибки намного меньше чем среднее отклонение признака MEDV.